StenosisMeasurement3D/qSlicerStenosisMeasurement3DModuleWidget.cxx

```diff
@@ -355,56 +428,294 @@ void qSlicerStenosisMeasurement3DModuleWidget::onTubePointEndInteraction(vtkObje
+  if (this->outputTableNode == node)
+  {
+    return;
+  }
+  this->outputTableNode = vtkMRMLTableNode::SafeDownCast(node);
+  if (d->ParameterNode)
+  {
+    d->ParameterNode->SetOutputTableNodeID(this->outputTableNode ? this->outputTableNode->GetID() : nullptr);
+  }
+}
+
+//-----------------------------------------------------------
+bool qSlicerStenosisMeasurement3DModuleWidget::setEditedNode(vtkMRMLNode* node,
+                                                  QString role /* = QString()*/,
+                                                  QString context /* = QString()*/)
+{
+  Q_D(qSlicerStenosisMeasurement3DModuleWidget);
+  Q_UNUSED(role);
+  Q_UNUSED(context);
+  
+  if (vtkMRMLStenosisMeasurement3DParameterNode::SafeDownCast(node))
+  {
+    d->parameterSetSelector->setCurrentNode(node);
+    return true;
+  }
+  return false;
+}
+
+//------------------------------------------------------------------------------
+void qSlicerStenosisMeasurement3DModuleWidget::setParameterNode(vtkMRMLNode* node)
+{
+  Q_D(qSlicerStenosisMeasurement3DModuleWidget);
+  if (this->inputFiducialNode)
+  {
+    // Disconnect the currently observed node.
+    this->inputFiducialNode->RemoveObserver(this->fiducialObservation);
+  }
+  if (this->inputShapeNode)
+  {
+    // Disconnect the currently observed node.
+    this->inputShapeNode->RemoveObserver(this->tubeObservation);
+  }
+  vtkMRMLStenosisMeasurement3DParameterNode* parameterNode = vtkMRMLStenosisMeasurement3DParameterNode::SafeDownCast(node);
+  d->ParameterNode = parameterNode;
+  d->EnableWorkspace();
+  this->updateWidgetFromMRML();
+}
+
+//----------------------------------------------------------------------------
+void qSlicerStenosisMeasurement3DModuleWidget::updateWidgetFromMRML()
+{
+  Q_D(qSlicerStenosisMeasurement3DModuleWidget);
+  if (!d->ParameterNode)
+  {
+    // Reset all UI widgets.
+    d->inputShapeSelector->setCurrentNode(nullptr);
+    d->inputFiducialSelector->setCurrentNode(nullptr);
+    d->inputSegmentSelector->setCurrentNode(nullptr);
+    d->inputSegmentSelector->setCurrentSegmentID(nullptr);
+    d->wallModelSelector->setCurrentNode(nullptr);
+    d->lumenModelSelector->setCurrentNode(nullptr);
+    d->outputTableSelector->setCurrentNode(nullptr);
+    d->wallResultLabel->clear();
+    d->lumenResultLabel->clear();
+    d->lesionResultLabel->clear();
+    d->stenosisResultLabel->clear();
+    d->lengthResultLabel->clear();
+    return;
+  }
+  
+  /*
+   * The parameter node will get in turn updated.
+   * Since we do not observe its changes because it's private,
+   * we will not end in infinite recursion.
+   * 
+   * The segmentation selector is special. The segmentID must be explicitly
+   * cleared if the segmentation is null.
+   */
+  d->inputShapeSelector->setCurrentNode(d->ParameterNode->GetInputShapeNode());
+  d->inputFiducialSelector->setCurrentNode(d->ParameterNode->GetInputFiducialNode());
+  bool wasBlocked = d->inputSegmentSelector->blockSignals(true);
```

Comment:
You can now use ` QSignalBlocker blocker1(d->inputSegmentSelector);` instead of two `blockSignals()` calls.

Reply:
Very handy.